src/test/scala/KinesisConsumerSpec.scala

```diff
@@ -0,0 +1,301 @@
+      commitableRecord.millisBehindLatest shouldBe recordsInput.getMillisBehindLatest
+    }
+  }
+
+  it should "not shutdown the worker if the stream is drained but has not failed" in new WorkerContext with TestData {
+    recordProcessor.initialize(initializationInput)
+    recordProcessor.processRecords(recordsInput)
+
+    eventually(verify(mockWorker, times(0)).shutdown())
+  }
+
+  it should "shutdown the worker if the stream terminates" in new WorkerContext(errorStream = true) with TestData {
+    recordProcessor.initialize(initializationInput)
+    recordProcessor.processRecords(recordsInput)
+
+    eventually(verify(mockWorker, times(1)).shutdown())
+  }
+
+  it should "not drop messages in case of back-pressure" in new WorkerContext with TestData {
+    // Create and send 10 records (to match buffer size)
+    for(i <- 1 to 10) {
+      val record = mock(classOf[Record])
+      when(record.getSequenceNumber).thenReturn(i.toString)
+      recordProcessor.processRecords(
+        recordsInput.withRecords(List(record).asJava))
+    }
+
+    // Should process all 10 messages
+    eventually(output.size shouldBe(10))
+
+    // Send a batch that exceeds the internal buffer size
+    for(i <- 1 to 50) {
+      val record = mock(classOf[Record])
+      when(record.getSequenceNumber).thenReturn(i.toString)
+      recordProcessor.processRecords(
+        recordsInput.withRecords(List(record).asJava))
+    }
+
+    // Should process all 50 messages
+    eventually(output.size shouldBe(60))
+
+    eventually(verify(mockWorker, times(0)).shutdown())
+  }
+
+  it should "not drop messages in case of back-pressure with multiple shard workers" in new WorkerContext with TestData {
+    recordProcessor.initialize(initializationInput)
+    recordProcessor2.initialize(initializationInput.withShardId("shard2"))
+
+    // Create and send 10 records (to match buffer size)
+    for(i <- 1 to 5) {
+      val record = mock(classOf[Record])
+      when(record.getSequenceNumber).thenReturn(i.toString)
+      recordProcessor.processRecords(
+        recordsInput.withRecords(List(record).asJava))
+      recordProcessor2.processRecords(
+        recordsInput.withRecords(List(record).asJava))
+    }
+
+    // Should process all 10 messages
+    eventually(output.size shouldBe(10))
+
+    // Each shard is assigned its own worker thread, so we get messages
+    // from each thread simultaneously.
+    def simulateWorkerThread(rp: IRecordProcessor): Future[Unit] = {
+      Future {
+        for (i <- 1 to 25) { // 10 is a buffer size
+          val record = mock(classOf[Record])
+          when(record.getSequenceNumber).thenReturn(i.toString)
+          rp.processRecords(recordsInput.withRecords(List(record).asJava))
+        }
+      }
+    }
+
+    simulateWorkerThread(recordProcessor)
+    simulateWorkerThread(recordProcessor2)
+
+    // Should process all 50 messages
+    eventually(output.size shouldBe(60))
```

Comment:
50 in the comment

Reply:
Right - it does 10 at first, then an additional 50, so 60 total. Comment could be made better